modules/core/shared/src/main/scala/io/circe/JsonObject.scala

```diff
@@ -275,31 +275,147 @@ object JsonObject {
    */
-  final val empty: JsonObject = new MapAndVectorJsonObject(Map.empty, Vector.empty)
+  final val empty: JsonObject = EmptyJsonObject
 
   /**
    * Construct a [[JsonObject]] with a single field.
    */
-  final def singleton(key: String, value: Json): JsonObject = new MapAndVectorJsonObject(Map((key, value)), Vector(key))
+  final def singleton(key: String, value: Json): JsonObject = new SingletonJsonObject(key, value)
 
   implicit final val showJsonObject: Show[JsonObject] = Show.fromToString
   implicit final val eqJsonObject: Eq[JsonObject] = Eq.fromUniversalEquals
 
+  /**
+   * An empty implementation of [[JsonObject]].
+   */
+  private[this] object EmptyJsonObject extends JsonObject {
+    override private[circe] def applyUnsafe(k: String): Json = null
+    override def apply(key: String): Option[Json] = None
+    override def contains(key: String): Boolean = false
+    override val size: Int = 0
+    override val isEmpty: Boolean = true
+    override val keys: Iterable[String] = Iterable.empty
+    override val values: Iterable[Json] = Iterable.empty
+    override val toMap: Map[String, Json] = Map.empty
+    override val toIterable: Iterable[(String, Json)] = Iterable.empty
+    override def add(k: String, j: Json): JsonObject = singleton(k, j)
+    override def +:(field: (String, Json)): JsonObject = singleton(field._1, field._2)
+    override def remove(key: String): JsonObject = this
+    override def traverse[F[_]](f: Json => F[Json])(implicit F: Applicative[F]): F[JsonObject] = F.pure(this)
+    override def mapValues(f: Json => Json): JsonObject = this
+
+    override private[circe] def appendToFolder(folder: Printer.PrintingFolder): Unit = {
+      val p = folder.pieces(folder.depth)
+      folder.writer.append(p.lBraces)
+      folder.writer.append(p.rBraces)
+    }
+  }
+
+  /**
+   * An implementation of [[JsonObject]] for objects containing a single field, with a lighter memory footprint
+   * than map based implementations.
+   */
+  private[this] final class SingletonJsonObject(field: String, value: Json) extends JsonObject {
+    override private[circe] def applyUnsafe(k: String): Json =
+      if (k == field) value else null
+
+    override def apply(key: String): Option[Json] =
+      Option.when(key == field)(value)
+
+    override def contains(key: String): Boolean =
+      key == field
+
+    override val size: Int =
+      1
+
+    override val isEmpty: Boolean =
+      false
+
+    override def keys: Iterable[String] =
+      Iterable.single(field)
+
+    override def values: Iterable[Json] =
+      Iterable.single(value)
+
+    override def toMap: Map[String, Json] =
+      Map(field -> value)
+
+    override def toIterable: Iterable[(String, Json)] =
+      Iterable.single(field -> value)
+
+    private def addOrReplace(k: String, j: Json, append: Boolean): JsonObject = {
+      if (k == field) new SingletonJsonObject(k, j)
+      else {
+        val fields = Map.from(Seq((field, value), (k, j)))
```

Comment:
Looks to me like these are equivalent, so lets use the simpler one.
```suggestion
        val fields = Map((field, value), (k, j))
```